src/elfpy/simulators.py

```diff
@@ -299,95 +321,92 @@ def run_simulation(self):
                     * self.market.init_share_price  # APR, apply return to starting price (no compounding)
                     # * self.market.share_price # APY, apply return to latest price (full compounding)
                 )
+                self.market.vault_apy = self.config.simulator.vault_apy[self.day]
             for daily_block_number in range(self.config.simulator.num_blocks_per_day):
                 self.daily_block_number = daily_block_number
                 last_block_in_sim = (self.day == self.config.simulator.num_trading_days - 1) and (
                     self.daily_block_number == self.config.simulator.num_blocks_per_day - 1
                 )
-                # if self.config.simulator.shuffle_users:
-                # self.rng.shuffle(np.asarray(self.agent_list))  # shuffle the agent order each block
                 self.collect_and_execute_trades(last_block_in_sim)
                 logging.debug("day = %d, daily_block_number = %d\n", self.day, self.daily_block_number)
                 self.market.log_market_step_string()
                 if not last_block_in_sim:
                     self.market.tick(self.step_size())
                     self.block_number += 1
         # simulation has ended
-        for agent in self.agent_list:
+        for agent in self.agents.values():
             agent.log_final_report()
         # fees_owed = self.market.calc_fees_owed()
 
     def collect_and_execute_trades(self, last_block_in_sim=False):
         """Get trades from the agent list, execute them, and update states"""
         if not isinstance(self.market, Market):
             raise ValueError("market not defined")
-
-        # TODO: BIG HACK to make this PR pass.  There is a bug with removing liquidity that
-        # doesn't account for open shorts.  Someone can remove all liquidity while there are open
-        # shorts, then when another user goes to close a short, there are no reserves, so the calc
-        # functions error due to division by zero.  Need to fix the accounting of reserves to
-        # account for open short positions.
-        if last_block_in_sim:  # get all of a agent's trades
-            # we are reversing order here to make sure that agent 0's remove_liquidity transaction
-            # is last
-            self.agent_list = self.agent_list[::-1]
-
-        for agent in self.agent_list:  # trade is different on the last block
+        # TODO: This is a HACK to prevent the initial LPer from rugging other agents.
+        # The initial LPer should be able to remove their liquidity and any open shorts can still be closed.
+        # But right now, if the LPer removes liquidity while shorts are open,
+        # then closing the shorts results in an error (share_reserves == 0).
```

Comment:
I think this is fine, but we can create a test to evaluate this behavior. once we do, we can remove this TODO. tracking issue #74